### PR TITLE
Implement type-safe API for random number generation

### PR DESCRIPTION
**Pull request by Xanewok**
Thanks for your work on the type-safe Windows CNG API! :heart: 
I found myself needing also RNG so I figured it'd be good to upload a patch here.

Tested the changes using Windows 10.0.18362.

### CODE REVIEW

File: src/random.rs
Comment (emgre) [inline src/random.rs]: Would be nice to not need this and document everything.

File: src/random.rs
Comment (emgre) [inline src/random.rs]: A full example like I did in the other module is needed here. It helps understanding how to use it.

File: src/random.rs
Comment (emgre) [inline src/random.rs]: I think we should make this (and the associated `RandomOptions`) private. The `BCRYPT_RNG_USE_ENTROPY_IN_BUFFER` does nothing starting from Windows 8 (and should probably never be used) and `BCRYPT_USE_SYSTEM_PREFERRED_RNG` should be hidden at creation.

File: src/random.rs
Comment (emgre) [inline src/random.rs]: Make this private.

File: src/random.rs
Comment (emgre) [inline src/random.rs]: Use camel case in enums (rustfmt will probably flag this).

File: src/random.rs
Comment (emgre) [inline src/random.rs]: Needs `///` instead of `//`.

File: src/random.rs
Comment (emgre) [inline src/random.rs]: I'm not really comfortable with that. Expect hides a `panic!` underneath. I prefer having explicit error handling and let user deal with errors.

File: src/random.rs
Comment (emgre) [inline src/random.rs]: Make this private.

**Review by emgre — CHANGES_REQUESTED** [review]
I would prefer having only two objects exposed publicly:
- `RandomAlgorithmId`: the same enum we have currently with an extra `SystemPrefered`
- `RandomNumberGenerator` (formerly `RandomAlgorithm`): a struct with two methods:
  - `fn open(id: RandomAlgorithmId) -> crate::Result<RandomNumberGenerator>`
  - `fn gen_random(&self, &mut [u8]) -> crate::Result<()>`

`RandomNumberGenerator::open` would take care of opening the algo provider or not, and also save the required flag internally.

File: src/random.rs
Comment (Xanewok) [reply src/random.rs]: It's a good convention to annotate with `#[deny(missing_docs)]` at the crate root for the libraries. I didn't want to enforce that here and so I forgot to remove the warning version but if you don't mind I'll put the deny version at the crate root instead.

File: src/random.rs
Comment (Xanewok) [reply src/random.rs]: `rustc` does the case linting and I wanted to retain the low level naming because it's not so intuitive (it's weird having a random generation algorithm referred to as 'RNG' where there are other options :smile: ) but you're right, I'll change that.

File: src/random.rs
Comment (Xanewok) [reply src/random.rs]: While I agree that this would simplify things and that this should probably be not used as you're saying, I don't think we get to be opinionated here as just a type-safe wrapper API.

After all, what is valid to do on the lower level should be possible to do here as well.

File: src/random.rs
Comment (Xanewok) [reply src/random.rs]: This can't be private as part of a public enum variant

**Comment by Xanewok** [comment]
Makes sense, thanks for the review!

I'll modify the public API and for now apply the thumbed up suggestions tomorrow.

File: src/random.rs
Comment (emgre) [reply src/random.rs]: I mean `BCRYPT_USE_SYSTEM_PREFERRED_RNG` will be usable through the `RandomAlgorithmId::SystemPrefered`, so the only thing left is the `BCRYPT_RNG_USE_ENTROPY_IN_BUFFER` which is not supported on Windows 7 and lower (extended support for Windows 7 ends in 10 days).

I guess we could simply add a `gen_random_with_entropy_in_buffer` method and mark it as deprecated, but tbh, I wouldn't mind just not supporting it.

File: src/random.rs
Comment (emgre) [reply src/random.rs]: My bad, I misread the code, I thought it was an `#!allow(missing_docs)]`.

But why do we need this? I thought clippy would warn on missing docs by default, no?

**Comment by emgre** [comment]
So I spent the evening listening to some [jazz fusion](https://youtu.be/P4aFrf86_o0) and ended implementing what I had in mind for this. You can have a look on my [`random`](https://github.com/emgre/win-crypto-ng/tree/random) branch. I think the public API of this is less scattered and easier to use, while still providing all the functionalities in a safe way. Let me know what you think about it 😉

**Comment by Xanewok** [comment]
Thanks for doing this! I left some review comments at https://github.com/emgre/win-crypto-ng/commit/38fce6ca2509204f8ee2a3c581efcd98509daaa5.

I think I prefer your `gen_random_with_entropy_in_buffer` approach in the end. Hopefully they won't add new options in the meantime :grinning:
One thing that I think we should change is the infallible `RandomNumberGenerator::system_preferred` constructor and this should be more or less good to go modulo git shenanigans :+1:

**Comment by Xanewok** [comment]
:wave: Did you have a chance to look at the review by any chance?

**Comment by Xanewok** [comment]
Squashed our changes together into a single co-authored commit and added small suggestions I mentioned (e.g. simple infallible default RNG) in the review.

I think this should be ready to go now :+1:

**Review by emgre — APPROVED** [review]

**Comment by emgre** [comment]
I simply added a top-level documentation and I'll merge now. Thanks!

**Pull request merged**
